streams/lms_metadata.py

```diff
@@ -0,0 +1,173 @@
+          ip = re.search(r'\((.*?)\)', ip_find.stdout).group(1)
+          port = re.search(r':(\d{4})', ip_find.stdout).group(1)
+          self.address = f"{ip}:{port}"
+
+        if self.debug:
+          print(f"Found LMS Server: {self.address}", flush=True)
+
+        player_json = {"id": 1,	"method": "slim.request",	"params": [self.player_name, ["players", "-", 100, "playerid"]]}
+        player_info = requests.get(f'http://{self.address}/jsonrpc.js', json=player_json, timeout=200)
+        player_load = json.loads(player_info.text)
+        if self.debug:
+          print(f"Loaded: {player_load}")
+        players = player_load['result']['players_loop']
+
+        for player in players:
+          connected = player['connected']
+          if connected and player['name'] == self.player_name:
+            if self.debug:
+              print(f'LMS Metadata connected to player: {self.player_name}')
+            connected = True
+          else:
+            time.sleep(0.1)
+      except Exception as e:
+        # When first creating an LMS stream, there can be random errors that will close the while loop
+        # typically when asking the player for info when there isn't a player linked to the stream yet
+        if self.debug:
+          print(f"Could not find server: {e}", flush=True)
+        time.sleep(self.meta_ref_rate)
+
+    while connected:
+      try:
+        track_json = {"id": 1, "method": "slim.request", "params": [ self.player_name, ["status", "-",100] ]}
+        track_info = requests.post(f'http://{self.address}/jsonrpc.js', json=track_json, timeout=1000)
+        track_load = json.loads(track_info.text)
+
+        track_id = track_load["result"]["playlist_loop"][0]["id"]
+        song_json = {"id":2,"method":"slim.request","params":[ self.player_name, ["songinfo","-",100,f"track_id:{track_id}"]]}
+        song_info = requests.post(f'http://{self.address}/jsonrpc.js', json=song_json, timeout=1000)
+        song_load = json.loads(song_info.text)
+
+        song_data = self.flatten(song_load['result']['songinfo_loop'])
+        track_data = self.flatten(track_load['result']['playlist_loop'])
+
+
+        # Previously this code was a series of try-catches and if-elses
+        # This new code should be much more efficient in the long term, as it should work even for untested stream types
+        if song_data.get('artist'):
+          self.meta.artist = song_data.get('artist')
+          self.meta.album = song_data.get('album')
+          self.meta.track = song_data.get('title')
+          self.meta.image_url = song_data.get('artwork_url')
+        else:
+          self.meta.artist = song_data.get('title')
+          self.meta.album = song_data.get('remote_title')
+          self.meta.track = track_data.get('title')
+          self.meta.image_url = 'static/imgs/lms.png'
+          if song_data.get('coverid'):
```

Comment:
You may want to split up the cover art handling from the other metadata. What happens if artist, album, and title are populated by the file is hosted locally? What happens if the artwork_url isnt provided but artist is?

Reply:
As far as I can tell this is not resolved